source/fuzz/transformation_expand_vector_reduction.cpp

```diff
@@ -0,0 +1,180 @@
+// Copyright (c) 2020 André Perez Maselco
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "source/fuzz/transformation_expand_vector_reduction.h"
+
+#include "source/fuzz/fuzzer_util.h"
+#include "source/fuzz/instruction_descriptor.h"
+
+namespace spvtools {
+namespace fuzz {
+
+TransformationExpandVectorReduction::TransformationExpandVectorReduction(
+    const spvtools::fuzz::protobufs::TransformationExpandVectorReduction&
+        message)
+    : message_(message) {}
+
+TransformationExpandVectorReduction::TransformationExpandVectorReduction(
+    const uint32_t instruction_result_id,
+    const std::vector<uint32_t>& fresh_ids) {
+  message_.set_instruction_result_id(instruction_result_id);
+  *message_.mutable_fresh_ids() =
+      google::protobuf::RepeatedField<google::protobuf::uint32>(
+          fresh_ids.begin(), fresh_ids.end());
+}
+
+bool TransformationExpandVectorReduction::IsApplicable(
+    opt::IRContext* ir_context,
+    const TransformationContext& transformation_context) const {
+  auto* instruction =
+      ir_context->get_def_use_mgr()->GetDef(message_.instruction_result_id());
+
+  // |instruction| must be defined.
+  if (!instruction) {
+    return false;
+  }
+
+  // |instruction| must be OpAny or OpAll.
+  if (instruction->opcode() != SpvOpAny && instruction->opcode() != SpvOpAll) {
+    return false;
+  }
+
+  // |instruction| must not be irrelevant.
+  if (transformation_context.GetFactManager()->IdIsIrrelevant(
+          instruction->result_id())) {
+    return false;
+  }
+
+  // |instruction| block must not be a dead block.
+  if (transformation_context.GetFactManager()->BlockIsDead(
```

Comment:
I think you can safely remove this check since you are inserting a synonym into the `instruction`'s block. Thus, if that block is dead then the instruction will be irrelevant and the condition above will fail.

Reply:
Alternatively, you can refactor this into an assertion.

Reply:
I got it, but the `fuzzerutil::CanMakeSynonymOf` function does not check if the `instruction` block is dead.

Reply:
I think `IdIsIrrelevant` should check that.

Reply:
Ow I got you. You're right.

Reply:
Done.